### ISSUE TITLE
Ressource limit when deploying

### ISSUE THREAD

**Issue by mboulhaj** [awslabs/amazon-sagemaker-examples#307]
I am following the sage maker documentation (mnist k-mean example) to train and deploy an ML model. My notebook instance is a ml.t2.medium, I have created a training job using a ml.m4.xlarge and I'm trying to deploy it. Given that my resource limit for ml.m4.xlarge is 1, I tried to deploy to a ml.t2.medium instance but I am getting this error 
```
ResourceLimitExceeded: An error occurred (ResourceLimitExceeded) when calling the CreateEndpoint operation: 
The account-level service limit 'ml.m4.xlarge for endpoint usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. 
Please contact AWS support to request an increase for this limit.
```
I don't get why I'm getting an error related to `ml.m4.xlage`while I am trying to deploy to an `ml.t2.medium`. Is it mandatory to use the same type of instance for training and deploying ? 

PS: to deploy I'm using this piece of code
```
kmeans_predictor = kmeans.deploy(initial_instance_count=1,
                                 instance_type='ml.t2.medium')
```

Thank you in advance

**Comment by djarpin** [comment]
Hi @mboulhaj - Sorry for the delayed response.  You do not have to deploy on the same instance type as you trained on.  If you are still experiencing problems, it may be best to open an AWS Support Case.  Unfortunately, I wasn't able to replicate your issue.  Thanks.

**Comment by djarpin** [comment]
Closing due to inactivity.  Feel free to re-open if you're still blocked.  Thanks.

**Comment by collinprather** [comment]
I am having this exact same issue, and others are as well on [this AWS forum](https://forums.aws.amazon.com/thread.jspa?threadID=274980). This issue may be worth re-opening.

**Comment by raghha** [comment]
I am still having the same issue even after raising the request to AWS support for increasing the resource limit.   After the service limit increase am able to create the ml.m4.xlarge instance from Sagemaker console, while deploying still getting the same error message
 ResourceLimitExceeded: An error occurred (ResourceLimitExceeded) when calling the CreateEndpoint operation: The account-level service limit 'ml.m4.xlarge for endpoint usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. Please contact AWS support to request an increase for this limit.

Kindly share your inputs to solve the above issue.

**Comment by neeld66** [comment]
Hi,
I am having the same issue, but with different instance types. I gave ml.m4.16xlarge instance type for deployment and ml.m4.xlarge for training. Still, while deploying, I am getting the resource limit exceeded error.

**Comment by evgenyslab** [comment]
Hello all,

I am experience this issue as well. I have a limit increase on ml.p4.xlarge SageMaker Training instances to 1, but when I try to deploy an endpoint using ml.m4.xlarge, I get the error that my service limit for endpoint usage is 0.

I opened a ticket with support but they are still trying to figure it out on the Sagemaker side. There is no option to request an endpoint-specific limit increase.

Further, if I try to use an ml.p4.xlarge endpoint - that is also rejected.

**Comment by yshuidie** [comment]
> Hello all,
> 
> I am experience this issue as well. I have a limit increase on ml.p4.xlarge SageMaker Training instances to 1, but when I try to deploy an endpoint using ml.m4.xlarge, I get the error that my service limit for endpoint usage is 0.
> 
> I opened a ticket with support but they are still trying to figure it out on the Sagemaker side. There is no option to request an endpoint-specific limit increase.
> 
> Further, if I try to use an ml.p4.xlarge endpoint - that is also rejected.


Thank you so much for this, please update us if there are further progress.

I am also following the sagemaker-examples and encountered the same issue by exactly the same code.
While trying, I also found such error messages:

> ClientError: An error occurred (ValidationException) when calling the CreateEndpointConfig operation: 1 validation error detected: Value 'ml.t3.medium' at 'productionVariants.1.member.instanceType' failed to satisfy constraint: Member must satisfy enum value set: [ml.p2.xlarge, ml.m5.4xlarge, ml.m4.16xlarge, ml.p3.16xlarge, ml.m5.large, ml.t2.xlarge, ml.p2.16xlarge, ml.c4.2xlarge, ml.c5.2xlarge, ml.c4.4xlarge, ml.c5.4xlarge, ml.c4.8xlarge, ml.c4.large, ml.c5.large, ml.c5.9xlarge, ml.c5.xlarge, ml.c4.xlarge, ml.t2.2xlarge, ml.t2.medium, ml.c5.18xlarge, ml.p3.2xlarge, ml.m5.xlarge, ml.m4.10xlarge, ml.t2.large, ml.m5.12xlarge, ml.m4.xlarge, ml.m5.24xlarge, ml.m4.2xlarge, ml.p2.8xlarge, ml.m5.2xlarge, ml.p3.8xlarge, ml.m4.4xlarge]

I tried 'ml.m5.4xlarge', 'ml.c4.2xlarge', and 'ml.m4.xlarge', they all reached limit. Not sure if all of them have the same issue though.

**Comment by savy2017** [comment]
> Hello all,
> 
> I am experience this issue as well. I have a limit increase on ml.p4.xlarge SageMaker Training instances to 1, but when I try to deploy an endpoint using ml.m4.xlarge, I get the error that my service limit for endpoint usage is 0.
> 
> I opened a ticket with support but they are still trying to figure it out on the Sagemaker side. There is no option to request an endpoint-specific limit increase.
> 
> Further, if I try to use an ml.p4.xlarge endpoint - that is also rejected.
> 
> Edit:
> After 2 weeks of talking to AWS support, the ml.m4.xlarge endpoint instance limit was raised - but I received no update regarding the change - had to manually test ability to deploy.

How to raise the request to AWS support for raising ml.m4.xlarge endpoint instance limit

**Comment by evgenyslab** [comment]
> > Hello all,
> > I am experience this issue as well. I have a limit increase on ml.p4.xlarge SageMaker Training instances to 1, but when I try to deploy an endpoint using ml.m4.xlarge, I get the error that my service limit for endpoint usage is 0.
> > I opened a ticket with support but they are still trying to figure it out on the Sagemaker side. There is no option to request an endpoint-specific limit increase.
> > Further, if I try to use an ml.p4.xlarge endpoint - that is also rejected.
> > Edit:
> > After 2 weeks of talking to AWS support, the ml.m4.xlarge endpoint instance limit was raised - but I received no update regarding the change - had to manually test ability to deploy.
> 
> How to raise the request to AWS support for raising ml.m4.xlarge endpoint instance limit

I opened a limit increase ticket for sagemaker and specified that I needed to increase instance limit on endpoint for ml.m4.xlarge. Took them about 2 weeks, and I only found out by testing every couple of days.

**Comment by mhalomari** [comment]
You have to request to increase:
Limit type: SageMaker
Region: [region]
Resource Type: SageMaker Notebooks
Limit: ml.p2.xlarge instances
New limit: [quantity]

**Comment by zenclive** [comment]
From the AWS console go to:

1. Top right corner of console and click on {your account name} > My Service Quotas
3. left nav "AWS Services"
4. Search for "ec2"
5. Select "Amazon Elastic Compute Cloud (Amazon EC2)"
6. Search for the size of the ec2 instance you want to raise quote on, eg "m4.large"
7. Click the result
8. Click the orange "Request Quota Increase"
9. In the "Change quota value" box enter the number required and press orange Request button. The request has now been made.
10. Click "Dashboard" from the left nav

I had to wait about 10-15 mins for an m4.large quota increase to take effect, but I only realised that by trying every few mins to run my Jupyter notebook script until it started running - the dashboard itself still says "Quota requested" but my script is now running.

Hope it helps someone out there :)

**Comment by mike-mi** [comment]
Very helpful zenclive, thanks.

**Comment by aishwat** [comment]
@mboulhaj  Even for `ml.t2.medium` It's giving the same error 
`An error occurred (ResourceLimitExceeded) when calling the CreateEndpoint operation: The account-level service limit 'ml.m4.xlarge for endpoint usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. Please contact AWS support to request an increase for this limit.`
After that, requested quota increase for m4.xlarge to 3, then launched a new notebook instance and tried, still the same.

Any help ?

**Comment by prabh27** [comment]
@zenclive @aishwat The resource allocation for On-Demand EC2 and Sagemaker notebooks is different. Even after getting the resource allocated on On-Demand EC2, you will not be able to use it for Sagemaker training.

To increase your Sagemaker notebook quota, create a new case on AWS Support (https://console.aws.amazon.com/support/home). 
Choose - Service Limit Increase
Limit Type - Sagemaker
Resource Type - Sagemaker training

**Comment by hh820** [comment]
Big THANKS to @zenclive, I followed his step-by-step direction to resolve the same issue described by many in this thread.

There are a couple of details I needed to modify - perhaps due to evolving nature of AWS services.

Step 5 - I could not find the instance name I was using "m4.xlarge".  Instead I selected "Running On-Demand Standard (A, C, D, H, I, M, R, T, Z) instances" from the drop-down-list.

Step 8 - In the "Change quota value" box I entered the value of 10.  Before doing so, I checked the default quota on my newly created account and noted it as 5.  The value in this box is absolute value - not an addition to the existing quota value (i.e., entering 2 here would have reduced my quota from 5 to 2) from what I understood by reading the fine prints just below this box.

I did not need to file another request, as noted by @prabh27 for Sagemaker notebooks.

As noted by @zenclive I needed to wait for 5-10 minutes to re-run the xgb.deploy code successfully even though  my support ticket status is still showing "Quota requested".

**Comment by NiklasA1** [comment]
I have the same issue. Requested additional resources still pending.

Very annoying, what's supposed to be a simple "Get Started" guide requires additional requests that are not mentioned there but you have to google to find out about.

**Comment by ixez** [comment]
I have the same issue that no matter what kind of instance type I choose to deploy, it still ends up with `ResourceLimitExceeded` error prompting the instance type I used for training.

**Comment by farhanhubble** [comment]
@ixez curious which instances you're using for training and deployment.

**Comment by tanyagupta** [comment]
`m1.m5.large` worked for me

**Comment by abhishekdbihani** [comment]
Just wanted to mention that I solved my issue by requesting an additional limit increase for SageMaker Hosting for the limit I wanted. This allowed me to deploy an endpoint for testing. Note that if you request only training limit increase it won't work for Hosting since apparently all the Sagemaker types seem to be independent. 

![sagemaker_hosting](https://user-images.githubusercontent.com/26830679/83930951-c827ba80-a75f-11ea-9f9b-9451a5638272.PNG)

**Comment by tanyagupta** [comment]
@abhishekdbihani did you use an instance of SageMaker Hosting for deployment, and an instance of SageMaker Training for training?

**Comment by fa-ahmad** [comment]
@tanyagupta the training and hosting(deployment) instances have to be requested separately even if it's the same instance, for example, ml.m4.xlarge, you need to request twice.

**Comment by tanyagupta** [comment]
@fa-ahmad thanks! So let's say you need to train and deploy. You would request SageMaker hosting for the instance (say ml.m4.xlarge) and then again request the SageMaker Training  for the same instance, ml.m4.xlarge?

**Comment by fa-ahmad** [comment]
@tanyagupta right. And you have to specify that you need one instance for training and once for hosting. Example1: You want to train a deep network and I want GPU acceleration. You would request twice, an `ml.p2.xlarge` for training and then an `ml.m4.xlarge` for hosting. Example 2: I do not need GPU and `ml.m4.xlarge` is adequate for training and hosting. I still need to raise two tickets, and request one instance of `ml.m4.xlarge` for training and another `ml.m4.xlarge` for hosting.

**Comment by tanyagupta** [comment]
Many thanks @fa-ahmad !

**Comment by avanijjain16** [comment]
> I am following the sage maker documentation (mnist k-mean example) to train and deploy an ML model. My notebook instance is a ml.t2.medium, I have created a training job using a ml.m4.xlarge and I'm trying to deploy it. Given that my resource limit for ml.m4.xlarge is 1, I tried to deploy to a ml.t2.medium instance but I am getting this error
> 
> ```
> ResourceLimitExceeded: An error occurred (ResourceLimitExceeded) when calling the CreateEndpoint operation: 
> The account-level service limit 'ml.m4.xlarge for endpoint usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. 
> Please contact AWS support to request an increase for this limit.
> ```
> 
> I don't get why I'm getting an error related to `ml.m4.xlage`while I am trying to deploy to an `ml.t2.medium`. Is it mandatory to use the same type of instance for training and deploying ?
> 
> PS: to deploy I'm using this piece of code
> 
> ```
> kmeans_predictor = kmeans.deploy(initial_instance_count=1,
>                                  instance_type='ml.t2.medium')
> ```
> 
> Thank you in advance

Any one able find solution for this. I had same issue. I got an error while training model I raised my request to 1 after raising case with AWS. Now I am getting error for deploying model. Instance I am using is ml.p2.xlarge. 
Very disappointed I waited for 2 days to get this fixed and again getting same error now for deployment. 

ResourceLimitExceeded: An error occurred (ResourceLimitExceeded) when calling the CreateEndpoint operation: The account-level service limit 'ml.p2.xlarge for endpoint usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. Please contact AWS support to request an increase for this limit.

trained model instance - 
stimator = PyTorch(entry_point="train.py",
                    source_dir="train",
                    role=role,
                    framework_version='0.4.0',
                    train_instance_count=1,
                    train_instance_type='ml.p2.xlarge',
                    hyperparameters={
                        'epochs': 10,
                        'hidden_dim': 200,
                    })

**Comment by tanyagupta** [comment]
TLDR; you need to request a training instance for training and deployment instance for deployment. I had the same issue
More in case you need it
https://medium.com/the-n00b-code-chronicles/avoiding-common-errors-during-sagemaker-deployment-598453e81cb1

**Comment by OrjwanZaafarani** [comment]
I solved the issue by deleting the endpoint configuration and re running the line

**Comment by scoropeza** [comment]
> I solved the issue by deleting the endpoint configuration and re running the line

This just saved me a lot of headaches!
